### DISCUSSION TITLE
Gossipsub max raw data size

### DISCUSSION THREAD

**Discussion by kalabukdima** [libp2p/rust-libp2p#5598] [Q&A] [answered]
Hello!
We're using gossipsub to broadcast batches of events, so I would like to know the exact limit of the message length that can be sent. There is a config option `max_transmit_size`, but it [is applied](https://github.com/libp2p/rust-libp2p/blob/master/protocols/gossipsub/src/behaviour.rs#L597) to the wrapper struct, not the raw data. How can I find the maximum number of bytes that can fit in the raw `data` vec?

**Comment by dariusc93** [comment]
I dont believe there is an explicit limit that is implied to the raw data itself. At least skimming over [gossipsub spec](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md) i dont see, so you would be free to increase the `max_transmit_size` to your use-case. Do keep in mind that increasing it also means that you would be accepting the message size from other peers up to that set limit.

**Comment by kalabukdima** [comment]
> I dont believe there is an explicit limit that is implied to the raw data itself. At least skimming over [gossipsub spec](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md) i dont see, so you would be free to increase the `max_transmit_size` to your use-case. Do keep in mind that increasing it also means that you would be accepting the message size from other peers up to that set limit.

I mean, there's no particular data size I'm trying to fit in the message, but I want to know how much data I can pass to the `publish` function without getting the `PublishError::MessageTooLarge` error. I could use something dumb like `config.max_transmit_size - 5000`, but why lose precious bytes if the limit may be calculated accurately?

I could try deriving it from the [`RawMessage`](https://github.com/libp2p/rust-libp2p/blob/master/protocols/gossipsub/src/types.rs#L99) struct size, but is it guaranteed not to change?

**Reply by dariusc93** [reply]
If I recall, you would only get `PublishError::MessageTooLarge` if the raw message protobuf exceeds the max transmit size. If it helps any, i was able to transmit 10MB message over gossip without hitting that error (not that you would want to send such a message anyway). 

> I could try deriving it from the [RawMessage](https://github.com/libp2p/rust-libp2p/blob/master/protocols/gossipsub/src/types.rs#L99) struct size, but is it guaranteed not to change?

The size there could possibly change. 

CC @jxs

**Reply by kalabukdima** [reply]
> you would only get PublishError::MessageTooLarge if the raw message protobuf exceeds the max transmit size

Exactly! And the raw message protobuf consists of the `data` passed in the `publish` function **plus** some surplus wrapping it. I can control the limit on the total size of the raw message (64k in my case), but I need to know the limit on the `data` vector, which can only be calculated as the `max_transmit_size` minus the size of the surplus.

If that surplus size would be exposed via public API and guaranteed to be updated whenever the `RawMessage` struct changes, that would solve my problem.

**Reply by kalabukdima** [reply]
@jxs could you take a look, please?

**Reply by jxs** [reply]
Hi Dzmitry and sorry for the delay! It's as Darius says, but after reading this I agree it probably would make sense to check the data to be published instead of the whole protobuf structure, would that help your case?

**Reply by kalabukdima** [reply]
Yes, if the `max_transmit_size` were imposed on the `data` passed to `publish` instead of the final wrapped message, it would help a lot. However, the receiver should also be able to get such a message, despite its total wire size exceeding the `max_transmit_size`.

**Reply by jxs** [reply]
would you like to submit a PR addressing it?

**Reply by kalabukdima** [reply]
I don't really understand what the receiver is supposed to do in this case.

Basically, there are two kinds of message sizes that can be limited.
- `user_message_size` — the size of the message passed into the `publish` function. That's the one I'd like to know the limit of.
- `wire_message_size` — the size of the actual message that gets transmitted through the network.

As I understand, there is a single limit on the `wire_message_size` on both sender and receiver ends. And we probably want the receiver to calculate the incoming message size before parsing, so it can only apply the limit on the `wire_message_size`, right?\
It brings us back to the original problem — if I know the maximum possible `user_message_size`, how do I calculate the maximum possible `wire_message_size` (or vice versa)?

**Reply by jxs** [reply]
you can't currently the maximum possible `wire_message_size` so what I am suggesting is that you submit a PR with this:
```diff
diff --git a/protocols/gossipsub/src/behaviour.rs b/protocols/gossipsub/src/behaviour.rs
index 16adb555a..a9b9c1942 100644
--- a/protocols/gossipsub/src/behaviour.rs
+++ b/protocols/gossipsub/src/behaviour.rs
@@ -583,6 +583,11 @@ where
             .data_transform
             .outbound_transform(&topic, data.clone())?;
 
+        // check that the size doesn't exceed the max transmission size
+        if transformed_data.len() > self.config.max_transmit_size() {
+            return Err(PublishError::MessageTooLarge);
+        }
+
         let raw_message = self.build_raw_message(topic, transformed_data)?;
 
         // calculate the message id from the un-transformed data
@@ -593,11 +598,6 @@ where
             topic: raw_message.topic.clone(),
         });
 
-        // check that the size doesn't exceed the max transmission size
-        if raw_message.raw_protobuf_len() > self.config.max_transmit_size() {
-            return Err(PublishError::MessageTooLarge);
-        }
-
         // Check the if the message has been published before
         if self.duplicate_cache.contains(&msg_id) {
             // This message has already been seen. We don't re-publish messages that have already
```

that would solve your problem right?

**Reply by kalabukdima** [reply]
I didn't realize the limit is not enforced on receiving or retransmitting messages. Made a PR: https://github.com/libp2p/rust-libp2p/pull/5642

**Reply by kalabukdima** [reply]
By the way, isn't it a vulnerability that any peer is allowed to push you an unlimited amount of data in the gossipsub messages?

**Comment by jxs** [comment, marked as answer]
Solved with https://github.com/libp2p/rust-libp2p/pull/5642
